### PR TITLE
perf(select): improve the performance when option mount and unmount

### PR DESCRIPTION
**Pull request by Ryan2128**
Please make sure these boxes are checked before submitting your PR, thank you!

* [ ] Make sure you follow Element's contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
* [ ] Make sure you are merging your commits to `dev` branch.
* [ ] Add some descriptions and refer to relative issues for your PR.

When options mount and unmount, a large number of array moves will be processed. At this time, `set` and `get` will be triggered many times, so use Map instead of Array.

#1512

### CODE REVIEW

**Comment by Ryan2128** [comment]
Hello, @siam-ese , you can check out this pr to test your demo.

**Comment by Ryan2128** [comment]
I think the remaining optimization will be virtual scrolling to reduce the time required for dom operations.

**Comment by siam-ese** [comment]
> Hello, @siam-ese , you can check out this pr to test your demo.

ok

File: packages/select/src/option.vue [hunk 1]
```diff
@@ -67,21 +67,20 @@ export default defineComponent({
...
 
     const vm = getCurrentInstance().proxy
-    select.onOptionCreate(vm)
+    select.onOptionCreate(vm as unknown as SelectOptionProxy)
 
     onBeforeUnmount(() => {
       const { selected } = select
       let selectedOptions = select.props.multiple ? selected : [selected]
-      let index = select.cachedOptions.indexOf(vm)
+      let isExist = select.cachedOptions.has(props.value)
```
Comment (HerringtonDarkholme) [inline packages/select/src/option.vue:75]: doesExist

File: packages/select/src/option.vue [hunk 2]
```diff
@@ -67,21 +67,20 @@ export default defineComponent({
...
       let selectedIndex = selectedOptions.findIndex(item => {
```
Comment (HerringtonDarkholme) [inline packages/select/src/option.vue:76]: `selectedOptions.some`

**Review by jw-foss — APPROVED** [review]

**Comment by JeremyWuuuuu** [comment]
LGTM

**Pull request merged**